lib/kanta/query.ex

```diff
@@ -134,6 +135,39 @@ defmodule Kanta.Query do
         )
       end
 
+      @doc """
+      Joins resource with another resource.
+
+      *Important!* The joining has to be defined by join_resource/2 function. Example:
+      ```
+      defp join_resource(query, :articles) do
+        query
+        |> join(:left, [user: u], _ in assoc(u, :articles), as: :article)
+      end
+      ```
+
+      First argument is a query and second argument is pattern-matched atom.
+
+      ## Examples
+
+          iex> Kanta.Accounts.UserQueries.with_join(:articles)
+          #Ecto.Query<from u0 in Kanta.Accounts.User, as: :user,
+            left_join: u1 in assoc(u0, :articles), as: :article>
+
+      """
+      @spec with_join(Ecto.Query.t(), atom()) :: Ecto.Query.t()
+      def with_join(query \\ base(), resource_name, opts \\ nil) when is_atom(resource_name) do
+        if has_named_binding?(query, resource_name) do
+          query
+        else
+          if is_nil(opts) do
+            join_resource(query, resource_name)
+          else
+            join_resource(query, resource_name, opts)
+          end
+        end
```

Comment:
Unfortunately, there is no function `without_named_binding/3`.

Reply:
Something is wrong then.
Checked docs and it's there since `3.9.0`(https://hexdocs.pm/ecto/3.9.0/Ecto.Query.html#with_named_binding/3).
Kanta's `mix.exs` specifies `~> 3.9` https://github.com/curiosum-dev/kanta/blob/main/mix.exs#L32C8-L32C23 .
Kanta's ecto lockfile is locked at 3.9.4.
So it should be there.

Reply:
My bad. You meant that you want a function that runs when there's no binding?

Reply:
If you made a typo it's called `with_named_binding/3`

Reply:
> You meant that you want a function that runs when there's no binding?
That's what I meant. Sorry for misleading you. 😄

Reply:
Well, `with_named_binding/3` does exactly that. It only runs the lambda when the binding doesn't exist. Take a look at the example code it looks exactly the same as your use case.

```elixir
if has_named_binding?(query, :comments) do
  query
else
  join(query, :left, c in assoc(p, :comments), as: :comments)
end
```

becomes

```elixir
with_named_binding(query, :comments, fn  query, binding ->
  join(query, :left, a in assoc(p, ^binding), as: ^binding)
end) 
```

in the example